rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,221 @@
+
+This section may be expanded into a standalone RFD in the future, to discuss this topic in more detail as TAG becomes more settled.
+
+### Alternative Designs
+
+Several alternative designs were considered for this feature, with one noteworthy option being the "managed roles" design, outlined below:
+
+1. The concept of "managed database roles" would be introduced. These roles would closely resemble the structure of the `db_permissions` field, and each managed database role would be stored as an individual resource within the backend.
+
+2. The database definition would be expanded to allow for referencing any number of these managed database roles.
+
+3. A database agent responsible for connecting to the database would synchronize the expected and actual database roles triggered by various events, mirroring the approach used for user permission synchronization.
+
+4. User roles, however, would remain unchanged. The existing mechanism for granting roles to automatically created users upon connection would still be employed.
+
+The primary distinction in this alternative design is the entity responsible for holding the database permissions. In the main design, it's the user, while in this alternative approach, it's a specifically designed role. This difference has implications for various aspects, including implementation, user experience, reliability, and security.
+
+The primary advantage of this alternative design is its ability to create database-level roles such as "sales," "accounting," or "devs" and grant permissions as needed. This functionality could be valuable on its own, even outside the context of Database Access, and might be integrated as part of a separate "Database Management" product or feature.
+
+However, this alternative design introduces added complexity, including the need for a new resource type, the requirement to establish the capability to reference this new resource from existing resources, and the need to coordinate the interaction between databases agents. Additionally, a mechanism for cleaning up old roles or "deactivating" them would be necessary, particularly concerning disconnected databases with outdated roles. This approach would also introduce a new task for the database agent related to "Database role management," which would be separate from the core proxy functionality.
+
+### Backward Compatibility
+
+The new `db_permissions` field should be automatically ignored by Teleport versions that don't support it.
+
+### Audit Events
+
+Since there are no specific audit events in place for the user creation feature, there is currently no requirement to extend any existing events. However, in the event that audit events are introduced, it would be prudent to contemplate enhancing these events by including a list of the applied permissions. This list would be a combination of pertinent entries in both `spec.allow.db_permissions` and `spec.deny.db_permissions`.
```

Comment:
Potentially outside the scope of this RFD.  But we should have audit events on connect and disconnect at a minimum.

Reply:
We do have connect and disconnect events, but we don't have 'user created', which is a distinct operation.

Reply:
Created and removed would be good to add as well